### PR TITLE
prometheus-knot-exporter: 2021-08-21 -> 3.3.1; python310Packages.libknot: init at 3.3.1

### PR DESCRIPTION
**Pull request by mweinelt**
Adds libknot, the Python bindings to interface with libknot and Knot DNS.

Updates prometheus-knot-exporter to the nic.cz upstream version.

Adapts the module and tests to the new version.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mweinelt** [comment]
Updated the module slightly. PTAL.

**Comment by marrobHD** [comment]
Result of `nixpkgs-review pr 258047` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>prometheus-knot-exporter</li>
    <li>prometheus-knot-exporter.dist</li>
    <li>python310Packages.libknot</li>
    <li>python310Packages.libknot.dist</li>
    <li>python311Packages.libknot</li>
    <li>python311Packages.libknot.dist</li>
  </ul>
</details>

File: pkgs/development/python-modules/libknot/default.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+  src = fetchPypi {
+    inherit pname version;
+    hash = "sha256-zoaDzqTj31wHsGCz/Qw/2J0bQ9CtkWqlSHb4y+JV8U8=";
+  };
+
+  postPatch = ''
+    substituteInPlace libknot/__init__.py \
+      --replace "libknot%s.dylib" "${knot-dns.out}/lib/libknot%s.dylib" \
+      --replace "libknot.so%s" "${knot-dns.out}/lib/libknot.so%s"
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/libknot/default.nix:25]:
```suggestion
      --replace "libknot%s.dylib" "${lib.getLib knot-dns}/lib/libknot%s.dylib" \
      --replace "libknot.so%s" "${lib.getLib knot-dns}/lib/libknot.so%s"
```
this should be robust for the future

File: pkgs/development/python-modules/libknot/default.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    hatchling
+  ];
+
+  pythonImportsCheck = [
+    "libknot"
+  ];
+
+  meta = with lib; {
+    description = "Python bindings for libknot";
+    homepage = "https://pypi.org/project/libknot/";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/libknot/default.nix:38]:
```suggestion
    homepage = "https://gitlab.nic.cz/knot/knot-dns/-/tree/master/python/libknot";
```
this is probably more helpful

File: pkgs/servers/monitoring/prometheus/knot-exporter.nix [hunk 3]
```diff
@@ -1,39 +1,41 @@
...
+    "knot_exporter"
+  ];
 
   passthru.tests = { inherit (nixosTests.prometheus-exporters) knot; };
 
   meta = with lib; {
-    homepage = "https://github.com/ghedo/knot_exporter";
-    description = " Prometheus exporter for Knot DNS";
+    description = "Prometheus exporter for Knot DNS";
+    homepage = "https://pypi.org/project/knot-exporter/";
```
Comment (SuperSandro2000) [inline pkgs/servers/monitoring/prometheus/knot-exporter.nix:36]:
```suggestion
    homepage = "https://gitlab.nic.cz/knot/knot-dns/-/tree/master/python/knot_exporter";
```

**Comment by SuperSandro2000** [comment]
@vcunat you might be interested in this

**Comment by mweinelt** [comment]
Updated to 3.3.2 and accounted for all the feedback.

**Review by SuperSandro2000 — APPROVED** [review]
planning to deploy this later but LGTM 🎉

**Review by Ma27 — COMMENTED** [review]
Deployed and it works fine for me :+1: 

Considering that a few metrics are named differently now I think it would make sense to mention this in the release notes though.

File: nixos/doc/manual/release-notes/rl-2311.section.md [hunk 4]
```diff
@@ -313,6 +313,8 @@
 
 - `rome` was removed because it is no longer maintained and is succeeded by `biome`.
 
+- The `prometheus-knot-exporter` was migrated to a version maintained by CZ.NIC. Various metric names have changed, so checking existing rules is recommended.
```
Comment (mweinelt) [inline nixos/doc/manual/release-notes/rl-2311.section.md:316]: Proposal.

**Review by Ma27 — APPROVED** [review]

**Pull request merged**